sdk/core/azure-core/src/main/java/com/azure/core/http/RequestOptions.java

```diff
@@ -0,0 +1,152 @@
+ *
+ * The JSON structure for the request is defined as follows:
+ * <pre>{@code
+ * {
+ *   "id": 0,
+ *   "category": {
+ *     "id": 0,
+ *     "name": "string"
+ *   },
+ *   "name": "doggie",
+ *   "photoUrls": [
+ *     "string"
+ *   ],
+ *   "tags": [
+ *     {
+ *       "id": 0,
+ *       "name": "string"
+ *     }
+ *   ],
+ *   "status": "available"
+ * }
+ * }</pre>
+ *
+ * To create a concrete request, Json builder provided in javax package is used here for demonstration. However, any
+ * other Json building library can be used to achieve similar results.
+ *
+ * {@codesnippet com.azure.core.experimental.http.requestoptions.createjsonrequest}
+ *
+ * Now, this string representation of the JSON request can be set as body of RequestOptions
+ *
+ * {@codesnippet com.azure.core.experimental.http.requestoptions.postrequest}
+ */
+public class RequestOptions {
+    private Consumer<HttpRequest> requestCallback = request -> { };
+    private ResponseStatusOption statusOption = ResponseStatusOption.DEFAULT;
+
+    /**
+     * Gets the request callback, applying all the configurations set on this RequestOptions.
+     * @return the request callback
+     */
+    public Consumer<HttpRequest> getRequestCallback() {
+        return this.requestCallback;
+    }
+
+    /**
+     * Gets under what conditions the operation raises an exception if the underlying response indicates a failure.
+     * @return the configured option
+     */
+    public ResponseStatusOption getStatusOption() {
+        return this.statusOption;
+    }
+
+    /**
+     * Adds a header to the HTTP request.
+     * @param header the header key
+     * @param value the header value
+     *
+     * @return the modified RequestOptions object
+     */
+    public RequestOptions addHeader(String header, String value) {
+        this.requestCallback = this.requestCallback.andThen(request ->
+            request.getHeaders().set(header, value));
+        return this;
+    }
+
+    /**
+     * Adds a query parameter to the request URL.
+     * @param parameterName the name of the query parameter
+     * @param value the value of the query parameter
+     * @return the modified RequestOptions object
+     */
+    public RequestOptions addQueryParam(String parameterName, String value) {
+        this.requestCallback = this.requestCallback.andThen(request -> {
+            String url = request.getUrl().toString();
+            request.setUrl(url + (url.contains("?") ? "&" : "?") + parameterName + "=" + value);
+        });
+        return this;
+    }
+
+    /**
+     * Adds a custom request callback to modify the HTTP request before it's sent by the HttpClient.
+     * The modifications made on a RequestOptions object is applied in order on the request.
+     *
+     * @param requestCallback the request callback
+     * @return the modified RequestOptions object
+     */
+    public RequestOptions addRequestCallback(Consumer<HttpRequest> requestCallback) {
```

Comment:
`requestCallback` may not convey the purpose of the callback. Other options:
- `addRequestModifier`
- `addRequestProcessor`
- `addRequestHandler`
- `addRequestTransformation`

Reply:
Sticking with `requestCallback` for now and may revisit in the future